src/main/java/org/opentripplanner/graph_builder/module/StreetLinkerModule.java

```diff
@@ -103,48 +103,89 @@ public void linkTransitStops(Graph graph, TransitModel transitModel) {
       if (tStop.hasPathways()) {
         continue;
       }
-      // check if stop is already linked, to allow multiple linking cycles
-      if (tStop.getDegreeOut() + tStop.getDegreeIn() > 0) {
+      // check if stop is already linked, to allow multiple idempotent linking cycles
+      if (tStop.isConnectedToGraph()) {
         continue;
       }
-      TraverseModeSet modes = new TraverseModeSet(TraverseMode.WALK);
 
-      if (OTPFeature.FlexRouting.isOn()) {
-        // If regular stops are used for flex trips, they also need to be connected to car routable
-        // street edges.
-        if (stopLocationsUsedForFlexTrips.contains(tStop.getStop())) {
-          modes = new TraverseModeSet(TraverseMode.WALK, TraverseMode.CAR);
-        }
+      // ordinarily stops only need to be accessible by foot
+      StopLinkType linkType = StopLinkType.WALK_ONLY;
+
+      if (
+        OTPFeature.FlexRouting.isOn() && stopLocationsUsedForFlexTrips.contains(tStop.getStop())
+      ) {
+        linkType = StopLinkType.WALK_AND_CAR;
       }
 
-      graph
-        .getLinker()
-        .linkVertexPermanently(
-          tStop,
-          modes,
-          LinkingDirection.BOTH_WAYS,
-          (vertex, streetVertex) ->
-            List.of(
-              StreetTransitStopLink.createStreetTransitStopLink(
-                (TransitStopVertex) vertex,
-                streetVertex
-              ),
-              StreetTransitStopLink.createStreetTransitStopLink(
-                streetVertex,
-                (TransitStopVertex) vertex
-              )
-            )
-        );
+      linkStopToStreetNetwork(tStop, linkType);
+
       //noinspection Convert2MethodRef
       progress.step(m -> LOG.info(m));
     }
     LOG.info(progress.completeMessage());
   }
 
+  /**
+   * Link a stop to the nearest "relevant" edges.
+   * <p>
+   * These are mostly walk edges but if a stop is used by a flex pattern it also needs to be
+   * car-accessible. Therefore, flex stops are ensured to be connected to the car-accessible
+   * edge. This may lead to several links being created.
+   */
+  private void linkStopToStreetNetwork(TransitStopVertex tStop, StopLinkType linkType) {
+    graph
+      .getLinker()
+      .linkVertexPermanently(
+        tStop,
+        WALK_ONLY,
+        LinkingDirection.BOTH_WAYS,
+        (walkVertex, walkStreetVertex) -> {
+          ArrayList<Edge> linkEdges = new ArrayList<>(
+            createStopLinkEdges((TransitStopVertex) walkVertex, walkStreetVertex)
+          );
+
+          // If regular stops or group stops are used for flex trips, they also need to be connected to car routable
+          // street edges. This does not apply to zones as street vertices store which zones they are part of.
+          if (
+            linkType == StopLinkType.WALK_AND_CAR && !walkStreetVertex.isConnectedToDriveableEdge()
+          ) {
+            graph
```

Comment:
Consider adding the edges returned here, not adding them inside the function (line 167).

Reply:
If you do this line 163-169 collapses as well.

Reply:
I'm not sure I follow you. `linkVertexPermanently` returns `void`. (That is indeed a bit strange and we don't really need to return anything here, but I'm just following the idiom.) I can explain a bit in a dev meeting,